File: src/main/kotlin/io/github/muindor/tcresearchsolver/solver/Inventory.kt
package io.github.muindor.tcresearchsolver.solver

const val DEFAULT_THRESHOLD = 50
const val BASE = 1.0
const val K = 1.0

class Inventory(val supply: Map<Aspect, Int>, val threshold: Int = DEFAULT_THRESHOLD)

fun makeInventory(entries: List<Pair<Aspect, Int>>, threshold: Int = DEFAULT_THRESHOLD): Inventory =
    Inventory(LinkedHashMap<Aspect, Int>().apply { entries.forEach { put(it.first, it.second) } }, threshold)

fun validateInventory(inv: Inventory) {
    if (inv.threshold <= 0) throw IllegalArgumentException("threshold must be > 0, got ${inv.threshold}")
    for ((a, n) in inv.supply) {
        if (n < 0) throw IllegalArgumentException("supply['$a'] must be a non-negative integer, got $n")
    }
}

private fun supplyOf(inv: Inventory, a: Aspect): Int = inv.supply[a] ?: 0

fun directPenalty(inv: Inventory, a: Aspect): Double {
    val s = supplyOf(inv, a)
    if (s >= inv.threshold) return 0.0
    if (s > 0) return BASE + K * (inv.threshold - s)
    return Double.POSITIVE_INFINITY
}

/**
 * obtainCost: min(direct penalty, craft-from-components via recipe).
 * Pass a fresh [cache] per solve (or per (inv, data) pair) for memoization (G7).
 * IMPORTANT: never reuse a cache across different (inv, data) pairs — stale costs would break
 * g_lb/globalMinObtain admissibility (mirrors the WeakMap<AspectData, WeakMap<Inventory, …>>
 * nesting in the TS oracle that enforces this automatically).
 */
fun obtainCost(inv: Inventory, data: AspectData, a: Aspect, cache: HashMap<Aspect, Double> = HashMap()): Double {
    fun rec(x: Aspect, stack: HashSet<Aspect>): Double {
        cache[x]?.let { return it }
        if (x in stack) return Double.POSITIVE_INFINITY // cycle guard (data is a DAG; defensive)
        var best = directPenalty(inv, x)
        val recipe = data.combinations[x]
        if (recipe != null) {
            stack.add(x)
            val craft = rec(recipe.first, stack) + rec(recipe.second, stack)
            stack.remove(x)
            if (craft < best) best = craft
        }
        cache[x] = best
        return best
    }
    return rec(a, HashSet())
}

fun globalMinObtain(inv: Inventory, data: AspectData, cache: HashMap<Aspect, Double> = HashMap()): Double {
    var min = Double.POSITIVE_INFINITY
    for (a in data.universe) {
        val c = obtainCost(inv, data, a, cache)
        if (c < min) min = c
    }
    return min
}

enum class Feasible { TRUE, FALSE, UNKNOWN }

class AllocBudget(val maxNodes: Int = 200_000)

class AllocationResult(
    val feasible: Feasible,
    val scarcityCost: Double,
    val craftOps: Int,
    val leafConsumption: Map<Aspect, Int>,
)

/** Reverse-topological order: a parent aspect appears before any of its recipe components. */
private fun reverseTopoOrder(data: AspectData, aspects: Iterable<Aspect>): List<Aspect> {
    val order = ArrayList<Aspect>()
    val seen = HashSet<Aspect>()
    fun visit(a: Aspect) {
        if (!seen.add(a)) return
        data.combinations[a]?.let { recipe ->
            visit(recipe.first)
            visit(recipe.second)
        }
        order.add(a) // components pushed before parent => reverse at the end
    }
    for (a in aspects) visit(a)
    order.reverse() // now parents precede components
    return order
}

private class AllocSub(val cost: Double, val craftOps: Int, val direct: LinkedHashMap<Aspect, Int>)

/**
 * [demand] MUST be an insertion-ordered map (e.g. [LinkedHashMap]) — the TS oracle uses an
 * ordered Map and tie-breaking in the DP depends on the traversal order of demand.keys (G2).
 */
fun allocate(
    inv: Inventory,
    data: AspectData,
    demand: Map<Aspect, Int>,
    budget: AllocBudget = AllocBudget(),
): AllocationResult {
    validateInventory(inv)

    // Aspects relevant to this demand = demand keys plus their full component closure.
    val order = reverseTopoOrder(data, demand.keys)   // parents precede components
    val idx = HashMap<Aspect, Int>().apply { order.forEachIndexed { i, a -> put(a, i) } }
    val n = order.size
    val penalty = DoubleArray(n) { directPenalty(inv, order[it]) }
    val supplyArr = IntArray(n) { inv.supply[order[it]] ?: 0 }
    val need0 = IntArray(n) { demand[order[it]] ?: 0 }

    // Memoized exact DP. rec(i, need) = optimal allocation of indices [i..n) given residual `need`
    // (only entries >= i are meaningful; crafts at j>=i push demand to indices > j by topo order).
    // Memo key = i + suffix-needs => identical subproblems are solved once.
    val memo = HashMap<String, Any>()   // AllocSub or INFEASIBLE sentinel
    val INFEASIBLE = Any()
    var nodes = 0
    var budgetExhausted = false

    fun rec(i: Int, need: IntArray): Any {
        if (budgetExhausted) return INFEASIBLE
        if (i == n) return AllocSub(0.0, 0, LinkedHashMap())
        // Memo key enumerates need[i..n) exactly like TS `need.slice(i).join(',')`
        val key = buildString {
            append(i); append('|')
            for (j in i until n) {
                if (j > i) append(',')
                append(need[j])
            }
        }
        memo[key]?.let { return it }
        if (++nodes > budget.maxNodes) {
            budgetExhausted = true
            return INFEASIBLE
        }

        val x = order[i]; val want = need[i]; val avail = supplyArr[i]; val pen = penalty[i]
        val recipe = data.combinations[x]
        val maxDirect = minOf(want, avail)

        var best: Any = INFEASIBLE

        var d = maxDirect
        while (d >= 0) {
            val c = want - d
            if (c > 0 && recipe == null) { d--; continue }   // primal cannot be crafted
            if (d > 0 && !pen.isFinite()) { d--; continue }  // unreachable (+Inf penalty)
            val need2 = need.copyOf()
            if (c > 0 && recipe != null) {
                // linkedSetOf dedupes a doubled component while mult still multiplies by 2 (G2 + plan note)
                for (comp in linkedSetOf(recipe.first, recipe.second)) {
                    val j = idx.getValue(comp)               // j > i by topo order
                    need2[j] = need2[j] + mult(data, comp, x) * c
                }
            }
            val sub = rec(i + 1, need2)
            if (budgetExhausted) return INFEASIBLE           // BLOCKER-fix: never memoize/return past exhaustion
            if (sub === INFEASIBLE) { d--; continue }
            sub as AllocSub
            val cost = (if (d > 0) d * pen else 0.0) + sub.cost
            val craftOps = c + sub.craftOps
            if (best === INFEASIBLE || cost < (best as AllocSub).cost ||
                (cost == best.cost && craftOps < best.craftOps)) {
                val direct = LinkedHashMap(sub.direct)
                if (d > 0) direct[x] = d
                best = AllocSub(cost, craftOps, direct)
            }
            d--
        }
        memo[key] = best
        return best
    }

    val result = rec(0, need0)

    // Budget exhaustion ALWAYS wins: an interrupted search is unproven, even if a feasible split was seen.
    if (budgetExhausted) {
        return AllocationResult(Feasible.UNKNOWN, Double.POSITIVE_INFINITY, 0, emptyMap())
    }
    if (result === INFEASIBLE) {
        return AllocationResult(Feasible.FALSE, Double.POSITIVE_INFINITY, 0, emptyMap())
    }
    result as AllocSub
    val leaf = LinkedHashMap<Aspect, Int>()
    for ((a, dd) in result.direct) if (dd > 0) leaf[a] = dd
    return AllocationResult(Feasible.TRUE, result.cost, result.craftOps, leaf)
}

File: reference/ts-solver/core/inventory.ts
import type { Aspect, AspectData } from '../data/aspects';
import { mult } from './aspectGraph';

export const DEFAULT_THRESHOLD = 50;
export const BASE = 1; // spec §4.1: base > 0
export const K = 1; // spec §4.1: k >= 0

export interface Inventory {
  /** Non-negative integer counts. Absent key => 0. */
  readonly supply: ReadonlyMap<Aspect, number>;
  /** Strictly > 0. */
  readonly threshold: number;
}

export function makeInventory(entries: ReadonlyArray<readonly [Aspect, number]>, threshold = DEFAULT_THRESHOLD): Inventory {
  return { supply: new Map(entries), threshold };
}

export function validateInventory(inv: Inventory): void {
  if (!(inv.threshold > 0) || !Number.isFinite(inv.threshold)) {
    throw new Error(`threshold must be > 0, got ${inv.threshold}`);
  }
  for (const [a, n] of inv.supply) {
    if (!Number.isInteger(n) || n < 0) {
      throw new Error(`supply['${a}'] must be a non-negative integer, got ${n}`);
    }
  }
}

function supplyOf(inv: Inventory, a: Aspect): number {
  return inv.supply.get(a) ?? 0;
}

export function directPenalty(inv: Inventory, _data: AspectData, a: Aspect): number {
  const s = supplyOf(inv, a);
  if (s >= inv.threshold) return 0;
  if (s > 0) return BASE + K * (inv.threshold - s);
  return Number.POSITIVE_INFINITY;
}

// Keyed on (AspectData, Inventory): obtainCost depends on BOTH the recipe graph and the supply,
// so omitting AspectData would return stale costs when an Inventory is reused with different data
// (and break g_lb/globalMinObtain admissibility in the solver). Nested WeakMaps auto-GC.
const obtainCache = new WeakMap<AspectData, WeakMap<Inventory, Map<Aspect, number>>>();

export function obtainCost(inv: Inventory, data: AspectData, a: Aspect): number {
  let byInv = obtainCache.get(data);
  if (!byInv) {
    byInv = new WeakMap();
    obtainCache.set(data, byInv);
  }
  let cache = byInv.get(inv);
  if (!cache) {
    cache = new Map();
    byInv.set(inv, cache);
  }
  return obtainRec(inv, data, a, cache, new Set());
}

function obtainRec(inv: Inventory, data: AspectData, a: Aspect, cache: Map<Aspect, number>, stack: Set<Aspect>): number {
  const memo = cache.get(a);
  if (memo !== undefined) return memo;
  if (stack.has(a)) return Number.POSITIVE_INFINITY; // cycle guard (data is a DAG; defensive)
  const direct = directPenalty(inv, data, a);
  let best = direct;
  const recipe = data.combinations.get(a);
  if (recipe) {
    stack.add(a);
    const craft = obtainRec(inv, data, recipe[0], cache, stack) + obtainRec(inv, data, recipe[1], cache, stack);
    stack.delete(a);
    if (craft < best) best = craft;
  }
  cache.set(a, best);
  return best;
}

export function globalMinObtain(inv: Inventory, data: AspectData): number {
  let min = Number.POSITIVE_INFINITY;
  for (const a of data.universe) {
    const c = obtainCost(inv, data, a);
    if (c < min) min = c;
  }
  return min;
}

// mult re-exported for the allocator below.
export { mult };

export interface AllocBudget {
  /** Max DFS nodes before returning feasible:'unknown'. Default 200_000. */
  readonly maxNodes: number;
}

export interface AllocationResult {
  readonly feasible: boolean | 'unknown';
  readonly scarcityCost: number;
  readonly craftOps: number;
  /** Actual per-aspect direct draws from supply (= direct[X]); subtracted by "Subtract used". */
  readonly leafConsumption: ReadonlyMap<Aspect, number>;
}

/** Reverse-topological order: an aspect appears before any of its recipe components. */
function reverseTopoOrder(data: AspectData, aspects: Iterable<Aspect>): Aspect[] {
  const order: Aspect[] = [];
  const seen = new Set<Aspect>();
  const visit = (a: Aspect): void => {
    if (seen.has(a)) return;
    seen.add(a);
    const recipe = data.combinations.get(a);
    if (recipe) for (const c of recipe) visit(c);
    order.push(a); // components pushed before parent => reverse at the end
  };
  for (const a of aspects) visit(a);
  order.reverse(); // now parents precede components
  return order;
}

const ALLOC_INFEASIBLE = Symbol('infeasible');
type AllocSub = { cost: number; craftOps: number; direct: Map<Aspect, number> };

export function allocate(
  inv: Inventory,
  data: AspectData,
  demand: ReadonlyMap<Aspect, number>,
  budget: AllocBudget = { maxNodes: 200_000 },
): AllocationResult {
  validateInventory(inv);

  // Aspects relevant to this demand = demand keys plus their full component closure.
  const order = reverseTopoOrder(data, demand.keys()); // parents precede components
  const idx = new Map<Aspect, number>(order.map((a, i) => [a, i]));
  const n = order.length;
  const penalty = order.map((a) => directPenalty(inv, data, a));
  const supplyArr = order.map((a) => inv.supply.get(a) ?? 0);
  const need0 = order.map((a) => demand.get(a) ?? 0);

  // Memoized exact DP. rec(i, need) = optimal allocation of indices [i..n) given residual `need`
  // (only entries >= i are meaningful; crafts at j>=i push demand to indices > j by topo order).
  // Memo key = i + suffix-needs => identical subproblems are solved once (spec §4.3 "memoized").
  const memo = new Map<string, AllocSub | typeof ALLOC_INFEASIBLE>();
  let nodes = 0;
  let budgetExhausted = false;

  const rec = (i: number, need: number[]): AllocSub | typeof ALLOC_INFEASIBLE => {
    if (budgetExhausted) return ALLOC_INFEASIBLE;
    if (i === n) return { cost: 0, craftOps: 0, direct: new Map() };
    const key = `${i}|${need.slice(i).join(',')}`;
    const cached = memo.get(key);
    if (cached !== undefined) return cached;
    if (++nodes > budget.maxNodes) { budgetExhausted = true; return ALLOC_INFEASIBLE; }

    const X = order[i]!;
    const want = need[i]!;
    const avail = supplyArr[i]!;
    const pen = penalty[i]!;
    const recipe = data.combinations.get(X);
    const maxDirect = Math.min(want, avail);

    let best: AllocSub | typeof ALLOC_INFEASIBLE = ALLOC_INFEASIBLE;
    for (let d = maxDirect; d >= 0; d--) {
      const c = want - d;
      if (c > 0 && !recipe) continue;                  // primal cannot be crafted
      if (d > 0 && !Number.isFinite(pen)) continue;    // unreachable (+Inf penalty)
      const need2 = need.slice();
      if (c > 0 && recipe) {
        for (const comp of new Set(recipe)) {
          const j = idx.get(comp)!;                    // j > i by topo order
          need2[j] = (need2[j] ?? 0) + mult(data, comp, X) * c;
        }
      }
      const sub = rec(i + 1, need2);
      if (budgetExhausted) return ALLOC_INFEASIBLE;    // BLOCKER-fix: never memoize/return past exhaustion
      if (sub === ALLOC_INFEASIBLE) continue;
      const cost = (d > 0 ? d * pen : 0) + sub.cost;
      const craftOps = c + sub.craftOps;
      if (best === ALLOC_INFEASIBLE || cost < best.cost || (cost === best.cost && craftOps < best.craftOps)) {
        const direct = new Map(sub.direct);
        if (d > 0) direct.set(X, d);
        best = { cost, craftOps, direct };
      }
    }
    memo.set(key, best);
    return best;
  };

  const result = rec(0, need0);

  // Budget exhaustion ALWAYS wins: an interrupted search is unproven, even if a feasible split was seen.
  if (budgetExhausted) {
    return { feasible: 'unknown', scarcityCost: Number.POSITIVE_INFINITY, craftOps: 0, leafConsumption: new Map() };
  }
  if (result === ALLOC_INFEASIBLE) {
    return { feasible: false, scarcityCost: Number.POSITIVE_INFINITY, craftOps: 0, leafConsumption: new Map() };
  }
  const leaf = new Map<Aspect, number>();
  for (const [a, dd] of result.direct) if (dd > 0) leaf.set(a, dd);
  return { feasible: true, scarcityCost: result.cost, craftOps: result.craftOps, leafConsumption: leaf };
}
